### PR TITLE
Adds missing module_specs for usages of _LazyModule

### LINKED ISSUE #15212: ValueError: transformers.models.auto.__spec__ is None. causing import errors
## Environment info

- `transformers` version: 4.15.0
- Platform: Colaboratory
- Python version: 3.7.12

### Who can help

@LysandreJik

## Information

Hello, this code was working last week but today I am getting a 'ValueError: transformers.models.auto.__spec__ is None' error which is causing errors when trying to import  other Libraries. I noted a similar issue #12904 but this has been resolved and closed last year.

Model I am using (Bert, XLNet ...): BERT

The problem arises when using: Transformers

My code:
```python
# Import all libraries
import pandas as pd
import numpy as np
import re


# Huggingface transformers
import transformers
from transformers import BertModel,BertTokenizer,AdamW, get_linear_schedule_with_warmup
print(transformers.__version__)
print(transformers.models.auto.__spec__)

import torch
from torch import nn 
from torch.utils.data import DataLoader,Dataset,RandomSampler, SequentialSampler
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_recall_fscore_support

import seaborn as sns
from pylab import rcParams
import matplotlib.pyplot as plt
from matplotlib import rc
%matplotlib inline

RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
torch.manual_seed(RANDOM_SEED)

import os
os.environ["CUDA_VISIBLE_DEVICES"]="0"
device = torch.device("cpu")
```

The Output:

```python
4.15.0
None
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-3e575e7ee253> in <module>()
     12 from torch import nn
     13 from torch.utils.data import DataLoader,Dataset,RandomSampler, SequentialSampler
---> 14 import pytorch_lightning as pl
     15 from pytorch_lightning.callbacks import ModelCheckpoint
     16 

10 frames
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/__init__.py in <module>()
     18 _PROJECT_ROOT = os.path.dirname(_PACKAGE_ROOT)
     19 
---> 20 from pytorch_lightning.callbacks import Callback  # noqa: E402
     21 from pytorch_lightning.core import LightningDataModule, LightningModule  # noqa: E402
     22 from pytorch_lightning.trainer import Trainer  # noqa: E402

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/callbacks/__init__.py in <module>()
     12 # See the License for the specific language governing permissions and
     13 # limitations under the License.
---> 14 from pytorch_lightning.callbacks.base import Callback
     15 from pytorch_lightning.callbacks.device_stats_monitor import DeviceStatsMonitor
     16 from pytorch_lightning.callbacks.early_stopping import EarlyStopping

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/callbacks/base.py in <module>()
     24 
     25 import pytorch_lightning as pl
---> 26 from pytorch_lightning.utilities.types import STEP_OUTPUT
     27 
     28 

/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/types.py in <module>()
     23 from torch.optim.lr_scheduler import _LRScheduler, ReduceLROnPlateau
     24 from torch.utils.data import DataLoader
---> 25 from torchmetrics import Metric
     26 
     27 _NUMBER = Union[int, float]

/usr/local/lib/python3.7/dist-packages/torchmetrics/__init__.py in <module>()
     12 _PROJECT_ROOT = os.path.dirname(_PACKAGE_ROOT)
     13 
---> 14 from torchmetrics import functional  # noqa: E402
     15 from torchmetrics.aggregation import CatMetric, MaxMetric, MeanMetric, MinMetric, SumMetric  # noqa: E402
     16 from torchmetrics.audio import (  # noqa: E402

/usr/local/lib/python3.7/dist-packages/torchmetrics/functional/__init__.py in <module>()
     12 # See the License for the specific language governing permissions and
     13 # limitations under the License.
---> 14 from torchmetrics.functional.audio.pit import permutation_invariant_training, pit, pit_permutate
     15 from torchmetrics.functional.audio.sdr import scale_invariant_signal_distortion_ratio, sdr, signal_distortion_ratio
     16 from torchmetrics.functional.audio.si_sdr import si_sdr

/usr/local/lib/python3.7/dist-packages/torchmetrics/functional/audio/__init__.py in <module>()
     12 # See the License for the specific language governing permissions and
     13 # limitations under the License.
---> 14 from torchmetrics.functional.audio.pit import permutation_invariant_training, pit, pit_permutate  # noqa: F401
     15 from torchmetrics.functional.audio.sdr import (  # noqa: F401
     16     scale_invariant_signal_distortion_ratio,

/usr/local/lib/python3.7/dist-packages/torchmetrics/functional/audio/pit.py in <module>()
     22 from torchmetrics.utilities import _future_warning
     23 from torchmetrics.utilities.checks import _check_same_shape
---> 24 from torchmetrics.utilities.imports import _SCIPY_AVAILABLE
     25 
     26 # _ps_dict: cache of permutations

/usr/local/lib/python3.7/dist-packages/torchmetrics/utilities/imports.py in <module>()
     90 _TQDM_AVAILABLE: bool = _module_available("tqdm")
     91 _TRANSFORMERS_AVAILABLE: bool = _module_available("transformers")
---> 92 _TRANSFORMERS_AUTO_AVAILABLE = _module_available("transformers.models.auto")
     93 _PESQ_AVAILABLE: bool = _module_available("pesq")
     94 _SACREBLEU_AVAILABLE: bool = _module_available("sacrebleu")

/usr/local/lib/python3.7/dist-packages/torchmetrics/utilities/imports.py in _module_available(module_path)
     34     """
     35     try:
---> 36         return find_spec(module_path) is not None
     37     except AttributeError:
     38         # Python 3.6

/usr/lib/python3.7/importlib/util.py in find_spec(name, package)
    112         else:
    113             if spec is None:
--> 114                 raise ValueError('{}.__spec__ is None'.format(name))
    115             return spec
    116 

ValueError: transformers.models.auto.__spec__ is None
```

## To reproduce

```python
import transformers
print(transformers.__version__)
print(transformers.models.auto.__spec__)

4.15.0
None
```

## Expected behavior

This code ran perfectly and all Libraries were imported last week. I made no changes to this code since but it produced the above error today.

### PR DESCRIPTION
**Pull request by jkuball**
# What does this PR do?

This PR adds a missing `__spec__` to the `_LazyModule` instances of all models.

When a module is missing a spec it could not be imported via `importlib.import_module`, which is what [torchmetrics does](https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/utilities/imports.py#L114), which is why I've found it in the first place.

Related to #13321 (pr)
Fixes #15212 (issue)

## Before submitting
- [x] ~This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).~
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] ~Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).~
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by eroehrig** [comment]
😍 exactly what i need

File: tests/test_configuration_auto.py [hunk 1]
```diff
@@ -26,6 +28,11 @@
 SAMPLE_ROBERTA_CONFIG = os.path.join(os.path.dirname(os.path.abspath(__file__)), "fixtures/dummy-config.json")
 
 
+def test_module_spec():
+    assert transformers.models.auto.__spec__ is not None
+    assert importlib.util.find_spec("transformers.models.auto") is not None
```
Comment (sgugger) [inline tests/test_configuration_auto.py:33]: Please use `self.assertIsNotNone` here.

**Review by sgugger — APPROVED** [review]
Thanks for fixing this. Note that it should be added in every `_LazyModule` but I can take care of that in a separate PR.

File: tests/test_configuration_auto.py (hunk 1 above)
Comment (jkuball) [reply tests/test_configuration_auto.py:33]: Since this is not part of a unittest.TestCase I can't

Comment (sgugger) [reply tests/test_configuration_auto.py:33]: It should be in the `AutoConfigTest`.

Comment (jkuball) [reply tests/test_configuration_auto.py:33]: I can and will do that! I just oriented my changes on #13321 which also uses the pure pytest variant.

Comment (sgugger) [reply tests/test_configuration_auto.py:33]: Yes, I didn't catch the test was not up to our usual standards in that one :-) Thanks for fixing here!

**Comment by jkuball** [comment]
I have just moved the test. I also have a local commit lying around that adds the module_spec to all occurences of the LazyModule, but I am not that sure on where to put the tests for that.

Looking at the `test_modeling_{{cookiecutter.lowercase_modelname}}.py` template I could add the test to both possibilities of the `{{cookiecutter.camelcase_modelname}}ModelTester` -- but I am not sure on how to import the model beforehand, since I am not that sure on how the Lazy Modules are working exactly.

Can I just add a `import transformers.models.{{cookiecutter.lowercase_modelname}}` to all model test files, even outside of `is_torch_available` or would this break things?

**Comment by sgugger** [comment]
Oh that's great! If you want to include that commit in this PR and rename it that would be great. Regarding tests, I don't think we need more than the test for the auto submodule and the test of transformers (since it's the same thing everywhere). To answer your question however, it does not need to be under a `is_torch_available()`.

To have the PR be perfectly complete, could you also add the change to the modeling template (if you don't have it in your commit already) as well as fix the test for transformers spec to be in a `unittest.TestCase`?

**Comment by jkuball** [comment]
Okay, I figured that with the tests, but I wasn't sure. In theory it wouldn't hurt, but yeah, it's probably okay to test the LazyModule once.

From my side this would now be mergeable, unless I made an error somewhere!

**Comment by sgugger** [comment]
Thank you so much for the wide fix! With the templates properly fixed, I don't think we will get new inits without the `module_spec` (if someone adds another new submodule, they will copy an existing init), so I think we're good.

**Pull request merged**